core/metrics_manager.js

```diff
@@ -0,0 +1,385 @@
+    top: absoluteTop,
+    left: absoluteLeft,
+  };
+};
+
+/**
+ * Gets the metrics for the visible workspace in either pixel or workspace
+ * coordinates. The visible workspace does not include the toolbox or flyout.
+ * @param {boolean=} opt_getWorkspaceCoordinates True to get the view metrics in
+ *     workspace coordinates, false to get them in pixel coordinates.
+ * @return {!Blockly.MetricsManager.ContainerRegion} The width, height, top and
+ *     left of the viewport in either workspace coordinates or pixel
+ *     coordinates.
+ * @public
+ */
+Blockly.MetricsManager.prototype.getViewMetrics = function(
+    opt_getWorkspaceCoordinates) {
+  var scale = opt_getWorkspaceCoordinates ? this.workspace_.scale : 1;
+  var svgMetrics = this.getSvgMetrics();
+  var toolboxMetrics = this.getToolboxMetrics();
+  var flyoutMetrics = this.getFlyoutMetrics(true);
+  var toolboxPosition = this.workspace_.toolboxPosition;
+
+  if (this.workspace_.getToolbox()) {
+    if (toolboxPosition == Blockly.TOOLBOX_AT_TOP ||
+        toolboxPosition == Blockly.TOOLBOX_AT_BOTTOM) {
+      svgMetrics.height -= toolboxMetrics.height;
+    } else if (toolboxPosition == Blockly.TOOLBOX_AT_LEFT ||
+        toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+      svgMetrics.width -= toolboxMetrics.width;
+    }
+  } else if (this.workspace_.getFlyout(true)) {
+    if (toolboxPosition == Blockly.TOOLBOX_AT_TOP ||
+        toolboxPosition == Blockly.TOOLBOX_AT_BOTTOM) {
+      svgMetrics.height -= flyoutMetrics.height;
+    } else if (toolboxPosition == Blockly.TOOLBOX_AT_LEFT ||
+        toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+      svgMetrics.width -= flyoutMetrics.width;
+    }
+  }
+  return {
+    height: svgMetrics.height / scale,
+    width: svgMetrics.width / scale,
+    top: -this.workspace_.scrollY / scale,
+    left: -this.workspace_.scrollX / scale,
+  };
+};
+
+/**
+ * Gets content metrics in either pixel or workspace coordinates.
+ *
+ * This can mean two things:
+ * If the workspace has a fixed width and height then the content
+ * area is rectangle around all the top bounded elements on the workspace
+ * (workspace comments and blocks).
+ *
+ * If the workspace does not have a fixed width and height then it is the
+ * metrics of the area that content can be placed. This area is computed by
+ * getting the rectangle around the top bounded elements on the workspace and
+ * adding padding to all sides.
+ * @param {boolean=} opt_getWorkspaceCoordinates True to get the content metrics
+ *     in workspace coordinates, false to get them in pixel coordinates.
+ * @return {!Blockly.MetricsManager.ContainerRegion} The
+ *     metrics for the content container.
+ * @public
+ */
+Blockly.MetricsManager.prototype.getContentMetrics = function(
+    opt_getWorkspaceCoordinates) {
+  var scale = opt_getWorkspaceCoordinates ? this.workspace_.scale : 1;
+  var viewSize = this.getViewMetrics(false);
+  var contentDimensions = this.getContentDimensions_(viewSize);
+
```

Comment:
Since we're allowing for getting only content metrics, I think we should refactor so that we only get view metrics if it's needed to compute content dimensions.

Reply:
done.